### PR TITLE
Recording an error may carry attributes

### PR DESCRIPTION
**Pull request by ktoso**
**Motivation:**
It can be used to attach more information bout the error, like its stacktrace and similar information.

This is also supported by the otel spec, so we're aligning closer with it here.

**Modifications:**

Change the recordError to accept optional span attributes

**Result:**

more powerful record error API

resolves https://github.com/apple/swift-distributed-tracing/issues/45

### CODE REVIEW

File: Sources/Tracing/Span.swift [hunk 1]
```diff
@@ -42,7 +42,7 @@ public protocol Span: AnyObject, _SwiftTracingSendableSpan {
     ///
     /// - Parameters:
     ///   - error: The error to be recorded.
```
Comment (slashmo) [inline Sources/Tracing/Span.swift:44]:
```suggestion
    /// - Parameters:
    ///   - error: The error to be recorded.
    ///   - attributes: Additional attributes describing the error.
```

File: Sources/Tracing/Span.swift [hunk 2]
```diff
@@ -97,6 +97,12 @@ extension Span {
     }
 }
 
+extension Span {
+    public func recordError(_ error: Error) {
```
Comment (slashmo) [inline Sources/Tracing/Span.swift:101]:
```suggestion
    /// Record a failure described by the given error.
    ///
    /// - Parameters:
    ///   - error: The error to be recorded.
    public func recordError(_ error: Error) {
```

File: Tests/TracingTests/TracerTests.swift [hunk 3]
```diff
@@ -253,6 +253,37 @@ final class TracerTests: XCTestCase {
...
+        tracer.withSpan("hello") { span in
+            span.recordError(errorToThrow, attributes: attrsForError)
+        }
+
+        XCTAssertTrue(endedSpan != nil)
+        XCTAssertEqual(endedSpan!.recordedErrors.count, 1)
+        let error = endedSpan!.recordedErrors.first!.0
+        XCTAssertEqual(error as! ExampleSpanError, errorToThrow)
+        let attrs = endedSpan!.recordedErrors.first!.1
+        XCTAssertEqual(attrs, attrsForError)
```
Comment (slashmo) [inline Tests/TracingTests/TracerTests.swift:283]: Do we want to use `XCTUnwrap` here instead of the force unwraps?

**Review by slashmo — APPROVED** [review]
Thanks for tackling this 👍

File: Tests/TracingTests/TracerTests.swift (hunk 3 above)
Comment (ktoso) [reply Tests/TracingTests/TracerTests.swift:283]: Meh, tbh I don't mind those `!`, thank you for the review!

**Pull request merged**